synapse/handlers/directory.py

```diff
@@ -307,15 +305,21 @@ def _update_canonical_alias(self, requester, user_id, room_id, room_alias):
             send_update = True
             content.pop("alias", "")
 
-        # Filter alt_aliases for the removed alias.
-        alt_aliases = content.pop("alt_aliases", None)
-        # If the aliases are not a list (or not found) do not attempt to modify
-        # the list.
-        if isinstance(alt_aliases, collections.Sequence):
-            send_update = True
-            alt_aliases = [alias for alias in alt_aliases if alias != alias_str]
-            if alt_aliases:
-                content["alt_aliases"] = alt_aliases
+        # Filter the alt_aliases property for the removed alias. Note that the
+        # value is not modified if alt_aliases is of an unexpected form.
+        try:
+            alt_aliases = content["alt_aliases"]
+            if isinstance(alt_aliases, (list, tuple)) and alias_str in alt_aliases:
```

Comment:
I wanted to add some tests for the case of `alt_aliases` being an unexpected type but couldn't figure out a way to insert "bad" state since it seems to go through the validation code that is added in `synapse.handlers.message`.

Reply:
we'd probably need to poke it in via the federation interface, which is altogether rather too much work tbh.